### DISCUSSION TITLE
Complicated API rule

### DISCUSSION THREAD

**Discussion by einarpersson** [pocketbase/pocketbase#3805] [Q&A] [answered]
I have the following collections
```
users
courseRegistrations
courseGroups
```

Each `courseRegistration` has a relation field to a single `user` and a single `courseGroup`.

I now want to create a rule for users, that allows a user to list / view all users in the same `courseGroup`.

I started writing this but quickly got stuck since I have so far only done simpler cases that are fairly straightforward and I lack the mental model / understanding for a case like this. Grateful for any advice on how to move forward.

**Comment by ganigeorgiev** [comment, marked as answer]
Hm, this unfortunately at the moment is a little more complicated.

If I understand correctly your current structure, you want a rule that will generate a pseudo-query like:

```sql
select users.*
from users
left join courseRegistrations as authcr on authcr.user = {:authId}
left join courseRegistrations as cr on cr.user = user.id
where authcr.courseGroup = cr.courseGroup
...
```
The problem in the above is the double JOIN of the same `courseRegistration` collection.
While we can perform a JOIN with the `@collection.*` rule, at the moment the filter rules doesn't support collection aliases and you can join the collection only once. This exists in the roadmap and I'll try to prioritize it for the next v0.20.0 release.

At the moment as a workaround, you can define a helper [View collection](https://pocketbase.io/docs/collections/#view-collection) that will have the same fields as the `courseRegistrations` and join that together with the original. For example the View collection can be named `courseRegistrationsView` with query like:

```sql
select id, user, courseGroup
from courseRegistrations
```

And then the `users` List API rule could look something like:

```js
@request.auth.id != "" &&
@collection.courseRegistrations.user ?= @request.auth.id &&
@collection.courseRegistrationsView.user ?= id
@collection.courseRegistrations.courseGroup ?= @collection.courseRegistrationsView.courseGroup 
```

But again, with v0.20.0 I'll consider prioritizing the `@collection.*` aliases to skip the extra view step. It is not something difficult, I just haven't thought much what syntax to use. If someone has suggestions free free to post them.

_edit: updated to use `?=` because otherwise it will require every joined record to match with the condition._

**Reply by ganigeorgiev** [reply]
Just an example how the above rule could look if we decide to use `@collection.someCollection:someAlias.*` syntax with the next release:

```js
@request.auth.id != "" &&
@collection.courseRegistrations.user ?= id &&
@collection.courseRegistrations:auth.user ?= @request.auth.id &&
@collection.courseRegistrations.courseGroup ?= @collection.courseRegistrations:auth.courseGroup 
```
But again, if someone has better idea, feel free to post it.

Eventually once we add support for back-relation references in the filters this can also look something like:
```js
@request.auth.id != "" &&
courseRegistrations(user).courseGroup ?= @request.auth.courseRegistrations(user).courseGroup
```
But back-relation references have some edge cases that need some more thinking before adding support for that.

**Reply by tonnax** [reply]
@ganigeorgiev I think this is one of the Pocketbase killer feature compare to others. I would really love to see **simpler** syntax in near future that is more easy to understand. If not, maybe a comprehensive documentation on practical use case examples. Something practical examples like (ie. in social media app): 
1. Only allow auth user to see all the "Post" collection's records where this record is created inside a group this auth user belong to
2. Only allow auth user to change the record that belong to them
3. Only allow auth user to create a new "Comment" collection record of a "Post" where this post is created inside a group this auth user belong to

also an idea: add functionally to return custom error code and error message from this API rules syntax as well if certain logic is met, no need to use Pocketbase Go framework in this case.

**Reply by VictorioBerra** [reply]
Stupid question @ganigeorgiev but is `id` a typo in the example?

`@collection.courseRegistrations.user ?= id &&` where is `id` coming from?

**Reply by ganigeorgiev** [reply]
@VictorioBerra This is a very old discussion that I've already forgot what was about but the expression `@collection.courseRegistrations.user ?= id` is valid and `id` refer to the `id` of the record in the collection where the rule is defined.

Note also that back-relations are already supported - https://pocketbase.io/docs/working-with-relations/#back-relations.

If you need further help, please open a new Q&A discussion but try to elaborate more clearly on your current collections structure and what actually you are trying to do.

**Reply by VictorioBerra** [reply]
@ganigeorgiev thanks for the quick reply, I am currently not trying to solve a problem I am just reading docs and validating my understanding. The docs reference this issue directly by link. The docs never mention a floating `Id` parameter so I was just very confused on what that was. Still trying to understand it to be honest.

**Reply by ganigeorgiev** [reply]
I don't understand what do you mean by "floating". The `id` is nothing special, it is just a field in a collection. You can perform the join with other fields (e.g. `relation`) if you want and it depends on your use case. For this particular case the `id` makes sense. To understand why read the original question and answer.

**Comment by verrol** [comment]
Since it seems that sub-queries and the `WITH` clause is supported in SQLite, can't the queries be written to use that instead? Or is using the alias feature just simpler?

**Reply by ganigeorgiev** [reply]
@verrol I don't understand what you are asking.

Note that both the alias and [Back-relations filtering](https://pocketbase.io/docs/working-with-relations/#back-relations) solutions mentioned in  https://github.com/pocketbase/pocketbase/discussions/3805#discussioncomment-7635080 are supported as of PocketBase v0.22 (_for the back-relation we ended up using `_via_` as separator instead of parenthesis per https://github.com/pocketbase/pocketbase/discussions/4417_).

If you need further help please open a new discussion and try to provide as much details as possible with your questions to understand the context and to avoid the back-and-forth.
